### PR TITLE
[XPU][TritonGEN] Revamp SIMD block memory access operations

### LINKED ISSUE #2750: Allow scalars as `triton_gen.simdblockread` and `triton_gen.simdblockwrite` type arguments
`triton_gen.simdblockread` and `triton_gen.simdblockwrite` represent scalar block accesses as length 1 vectors. This may lead to issues in the backend. As an improvement and for future proofing, allow scalars as `triton_gen.simdblockread` result types and `triton_gen.simdblockwrite` argument types.

### PR DESCRIPTION
**Pull request by victor-eds**
Revamp `tritongen.simdblock[read|write]` operations:

- Rename to `tritongen.sub_group_block_[read|write]`
- Implement type verification in signature
- Represent scalar block memory accesses with a scalar type instead of `vector<1xty>`
- Revamp ASM format

Closes #2750

### CODE REVIEW

File: test/TritonGEN/tritongen-invalid.mlir [hunk 1]
```diff
@@ -428,19 +428,3 @@ llvm.func @matrix_2Dblockprefetch(%ptr : !llvm.ptr, %base_width : i32, %base_hei
...
-  llvm.return
-}
-
-// -----
-
-llvm.func @triton_gen.simdblockwrite(%ptr: !llvm.ptr<3>, %val: vector<64xi16>) {
-  // expected-error @+1 {{'triton_gen.simdblockwrite' op unsupported vector type}}
-  triton_gen.simdblockwrite %ptr, %val : (!llvm.ptr<3>, vector<64xi16>)
-  llvm.return
-}
```
Comment (victor-eds) [inline test/TritonGEN/tritongen-invalid.mlir:446]: No need to check as this is encoded in the MLIR typesystem

**Review by sommerlukas — APPROVED** [review]

File: third_party/intel/include/Dialect/TritonGEN/IR/TritonGENOps.td [hunk 2]
```diff
@@ -314,46 +314,97 @@ def TritonGEN_Matrix2DBlockPrefetchOp : TritonGEN_Op<"2Dblockprefetch">,
...
+    ```
+    ptr[sub_group_local_id + sub_group_size]
+    ```
+    etc.
+
+    `ptr` must be aligned to the size of the element type of `res`.
+
+     Example:
+     ```mlir
+     %0 = triton_gen.sub_group_block_write %ptr, %val : !llvm.ptr<1>, vector<4xi32>
```
Comment (mfrancepillois) [inline third_party/intel/include/Dialect/TritonGEN/IR/TritonGENOps.td:395]: Nit: Perhaps you could mention that the subgroup size is 128 in this example.

**Review by mfrancepillois — APPROVED** [review]
LGTM

File: third_party/intel/include/Dialect/TritonGEN/IR/TritonGENOps.td (hunk 2 above)
Comment (victor-eds) [reply third_party/intel/include/Dialect/TritonGEN/IR/TritonGENOps.td:395]: Sub-group size is not encoded in the operation itself. It defines the semantics of the operation, as `vector_length*sub_group_size` elements will be accessed, but it is not encoded.

File: third_party/intel/include/Dialect/TritonGEN/IR/TritonGENOps.td [hunk 3]
```diff
@@ -314,46 +314,97 @@ def TritonGEN_Matrix2DBlockPrefetchOp : TritonGEN_Op<"2Dblockprefetch">,
...
+		              ".getAddressSpace() == " #
+			      "static_cast<unsigned>(kCrossWorkgroup)">,
+		    CPred<"::llvm::cast<::mlir::LLVM::LLVMPointerType>($_self)" #
+		              ".getAddressSpace() == " #
+			      "static_cast<unsigned>(kWorkgroup)">]>]>,
+           "LLVM pointer in local or global OpenCL address space",
+           "::mlir::LLVM::LLVMPointerType">;
+
+def TritonGEN_SubGroupBlockReadOp: TritonGEN_Op<"sub_group_block_read"> {
+  let summary = "Sub-group block read.";
```
Comment (victor-eds) [inline third_party/intel/include/Dialect/TritonGEN/IR/TritonGENOps.td:342]: Reminder before merge: use spaces instead of tabs for indentation.

**Review by whitneywhtsang — APPROVED** [review]

**Review by chengjunlu — APPROVED** [review]

**Pull request merged**